### PR TITLE
tfm: Add option and support for logging using a shared UART instance

### PR DESCRIPTION
**Pull request by joerchan**
Add option and support for configuring the UART0 instance as a shared peripheral between secure and non-secure.
The secure will only use it for logging during boot and on fatal errors.

NCSDK-17221

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

### CODE REVIEW

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: subsys/nonsecure/Kconfig [hunk 1]
```diff
@@ -12,7 +12,8 @@ menu "TF-M UART pins"
 
 config TFM_SECURE_UART1
 	bool
-	select NRF_UARTE1_SECURE
+	select NRF_UARTE0_SECURE if TFM_LOG_SHARE_UART
```
[the remark this answers is not in the document]
Comment (joerchan) [reply subsys/nonsecure/Kconfig:15]: Yeah, but the naming of this option runs deep. It's a zephyr Kconfig option that maps to the TF-M option "SECURE_UART1".
It is hardcoded all over TF-M.

Not sure what is the most elegant solution to this.

File: modules/tfm/tfm/boards/board/RTE_Device.h [hunk 2]
```diff
@@ -10,26 +10,30 @@
 #include <autoconf.h>
 
 /* Configuration settings for Driver_USART0. */
+#if (defined(CONFIG_TFM_SECURE_UART1) && defined(CONFIG_TFM_LOG_SHARE_UART)) \
```
Comment (mswarowsky) [inline modules/tfm/tfm/boards/board/RTE_Device.h:13]: Shouldn't that be `CONFIG_TFM_SECURE_UART0` here?
Or then at lest leave a comment why UART1

File: modules/trusted-firmware-m/Kconfig [hunk 3]
```diff
@@ -26,18 +26,45 @@ config TFM_REGRESSION_NS
 	bool
 	select NRF_TIMER0_SECURE
 
-config TFM_SECURE_UART1
+config TFM_SECURE_UART
 	bool "TF-M configure UART1 as secure peripheral"
```
Comment (lemrey) [inline modules/trusted-firmware-m/Kconfig:30]: Does this line need an update too?

File: modules/trusted-firmware-m/Kconfig [hunk 4]
```diff
@@ -26,18 +26,45 @@ config TFM_REGRESSION_NS
...
+	  up, and when handling a fatal error.
+
+choice TFM_SECURE_UART_INSTANCE
+	prompt "TF-M logging UART instance"
+	default TFM_SECURE_UART1
+
+config TFM_SECURE_UART0
+	bool "TF-M logging will use UART0 instance"
+	select NRF_UARTE0_SECURE
+	select TFM_SECURE_UART_SHARE_INSTANCE if NRFX_UARTE0 || HAS_HW_NRF_UARTE0
```
Comment (lemrey) [inline modules/trusted-firmware-m/Kconfig:57]: why the `||` here? the two are not equivalent

Comment (lemrey) [reply modules/trusted-firmware-m/Kconfig:57]: Shouldn't only `NRFX_UARTE0` be sufficient?

Comment (joerchan) [reply modules/trusted-firmware-m/Kconfig:57]: `NRFX_UARTE0` is not enabled even though zephyr is using it for output.
`HAS_HW_NRF_UARTE1` is not enabled when NRF_MODEM_LIB_TRACE_BACKEND_UART.
So my conclusion is that they are not equivalent.

**Review by eivindj-nordic — COMMENTED** [review]
I tried this locally with the nRF9160 download sample with TF-M and application logging on UART0 and modem traces on UART1 by adding 
```
CONFIG_TFM_LOG_LEVEL_SILENCE=n
CONFIG_TFM_SECURE_UART0=y
CONFIG_NRF_MODEM_LIB_TRACE=y
```
I had to change one line in the TF-M repo to make it work:
`modules/tee/tf-m/trusted-firmware-m/config/check_config.cmake:54` to 
`tfm_invalid_config(CONFIG_TFM_LOG_SHARE_UART AND NOT (SECURE_UART1 OR SECURE_UART0))`. Not sure if this is the correct fix, though.  

I also tested changing the chosen zephyr console and TF-M logs to uart1, and that also works (without modem traces).

File: lib/nrf_modem_lib/trace_backends/uart/Kconfig [hunk 5]
```diff
@@ -12,8 +12,6 @@ config NRF_MODEM_LIB_TRACE_BACKEND_UART
 	# Modem tracing over UART use the UARTE1 as a dedicated peripheral.
 	# This enables the UARTE1 peripheral and includes nrfx UARTE driver.
 	select NRFX_UARTE1
-	# TF-M use UARTE1 as log output, disable TF-M logging to avoid conflict.
```
[the remark this answers is not in the document]
Comment (joerchan) [reply lib/nrf_modem_lib/trace_backends/uart/Kconfig:15]: Most likely you would want to share the UART on UART1 in that case. Maybe best to add an imply TFM_SECURE_UART0 here?

**Comment by joerchan** [comment]
> I tried this locally with the nRF9160 download sample with TF-M and application logging on UART0 and modem traces on UART1 by adding
> 
> ```
> CONFIG_TFM_LOG_LEVEL_SILENCE=n
> CONFIG_TFM_SECURE_UART0=y
> CONFIG_NRF_MODEM_LIB_TRACE=y
> ```
> 
> I had to change one line in the TF-M repo to make it work: `modules/tee/tf-m/trusted-firmware-m/config/check_config.cmake:54` to `tfm_invalid_config(CONFIG_TFM_LOG_SHARE_UART AND NOT (SECURE_UART1 OR SECURE_UART0))`. Not sure if this is the correct fix, though.
> 
> I also tested changing the chosen zephyr console and TF-M logs to uart1, and that also works (without modem traces).

I cannot see why you would have to do that, I tried it myself and it builds fine here.
```
west build -b nrf9160dk_nrf9160_ns samples/nrf9160/download -- \
-DCONFIG_TFM_LOG_LEVEL_SILENCE=n \
-DCONFIG_TFM_SECURE_UART0=y \
-DCONFIG_NRF_MODEM_LIB_TRACE=y
```

File: subsys/nonsecure/Kconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply subsys/nonsecure/Kconfig]: I realized that TFM_SECURE_UART is defined in sdk-nrf, so I've done refactoring of the configurations that hides most of the confusion away from the user of the system.
But still SECURE_UART1 is the TF-M cmake option that enables secure UART. And in the nordic platform there is an option to select the UART instance.

File: modules/tfm/tfm/boards/board/RTE_Device.h (hunk 2 above)
Comment (joerchan) [reply modules/tfm/tfm/boards/board/RTE_Device.h]: Refactored the option, that should clear it up here.

**Comment by eivindj-nordic** [comment]
> I cannot see why you would have to do that, I tried it myself and it builds fine here.
I tried it again now, and it works fine here too. Sorry for the confusion!

File: modules/trusted-firmware-m/Kconfig [hunk 6]
```diff
@@ -26,17 +26,45 @@ config TFM_REGRESSION_NS
...
-	depends on SOC_NRF5340_CPUAPP || SOC_NRF9160
 	help
-	  Configure the UART1 peripheral as secure for TF-M logging.
-	  This makes the UART1 peripheral unavailable to the non-secure
+	  Configure the UART instance as a secure peripheral for TF-M logging.
+	  This makes the UART instance unavailable to the non-secure unless
+	  TFM_SECURE_UART_SHARE_INSTANCE is enabled.
 	  application. When this option is selected the device tree node for
-	  UART1 needs to be disabled for the non-secure application.
+	  the UART instance needs to be disabled for the non-secure application.
```
Comment (SebastianBoe) [inline modules/trusted-firmware-m/Kconfig:37]: Delete "application. " on line 36

**Review by SebastianBoe — APPROVED** [review]

File: lib/nrf_modem_lib/trace_backends/uart/Kconfig (hunk 5 above)
[the remark this answers is not in the document]
Comment (joerchan) [reply lib/nrf_modem_lib/trace_backends/uart/Kconfig:15]: Couldn't do this, imply cannot be used for select symbols.

**Review by eivindj-nordic — APPROVED** [review]

**Review by mswarowsky — APPROVED** [review]
lgtm

**Pull request merged**